### PR TITLE
Support reservation in thread pool

### PR DESCRIPTION
**Pull request by littlepig2013**
Summary:

Add `ReserveThreads` and `ReleaseThreads` functions in thread pool to support reservation in for a specific thread pool.  With this feature, a thread will be blocked if the number of waiting threads (noted by `num_waiting_threads_`) equals the number of reserved threads (noted by `reserved_threads_`), normally `reserved_threads_` is upper bounded by `num_waiting_threads_`; in rare cases (e.g. `SetBackgroundThreadsInternal` is called when some threads are already reserved), `num_waiting_threads_` can be less than `reserved_threads`. 

Test Plan:
Add `ReserveThreads` unit test in `env_test`. Update the unit test `SimpleColumnFamilyInfoTest` in `thread_list_test` with adding `ReserveThreads` related assertions.

### CODE REVIEW

File: util/threadpool_imp.cc [hunk 1]
```diff
@@ -100,6 +102,46 @@ struct ThreadPoolImpl::Impl {
...
+  int ReserveThreads(int threads_to_be_reserved) {
+    std::unique_lock<std::mutex> lock(mu_);
+    int curr_reserved_threads =
+        reserved_threads_.load(std::memory_order_relaxed);
+    int reserved_threads_in_success = threads_to_be_reserved;
+    // Assume that at least one thread is left in threadpool. For example, when
+    // we use ReserveThreads in subcompaction, the thread that calls
+    // subcompaction should be taken into account in the threadpool
+    if (curr_reserved_threads + threads_to_be_reserved + 1 >
+        total_threads_limit_) {
```
Comment (ajkr) [inline util/threadpool_imp.cc:115]: `total_threads_limit_` seems too high as some of those threads might be doing work and therefore should not be reserved. I think there should be a counter, `num_waiting_`, for threads in `bgsignal_.wait(lock)`. Then `num_waiting_` is the upper bound on how many threads can be reserved. Also you don't need to worry about `+ 1` because if this executes inside a thread-pool managed thread, that thread will naturally be excluded from `num_waiting_` so will be non-reservable.

File: util/threadpool_imp.cc [hunk 2]
```diff
@@ -320,6 +366,8 @@ void ThreadPoolImpl::Impl::SetBackgroundThreadsInternal(int num,
   if (num > total_threads_limit_ ||
       (num < total_threads_limit_ && allow_reduce)) {
     total_threads_limit_ = std::max(0, num);
+    // Reset the number of reserved threads when total_threads_limit_ changes
+    reserved_threads_.store(0, std::memory_order_relaxed);
```
Comment (ajkr) [inline util/threadpool_imp.cc:370]: Is it needed? User may call `SetBackgroundThreads()` at any time and RocksDB has no way to know its reservations have been invalidated.

**Review by ajkr — CHANGES_REQUESTED** [review]
Thanks for the PR, had a few comments.

Let's mention the Public API change in HISTORY.md so users know to update their `Env` implementation.

I think total_threads_limit_ is always accessed with ThreadPoolImpl mutex held, in which case we don't need to use atomic operations.

File: include/rocksdb/threadpool.h [hunk 3]
```diff
@@ -49,6 +49,15 @@ class ThreadPool {
...
   virtual void SubmitJob(std::function<void()>&&) = 0;
+
+  // Reserve available background threads. This function does not ensure
+  // so many threads can be reserved, instead it will return the number of
+  // threads that can be reserved against the desired one. In other words,
+  // the number of available threads could be less than the input.
+  virtual int ReserveThreads(int threads_to_be_reserved) = 0;
+
+  // Release a specific number of reserved threads
+  virtual int ReleaseThreads(int threads_to_be_released) = 0;
```
[the remark this answers is not in the document]
Comment (littlepig2013) [reply include/rocksdb/threadpool.h:60]: Yes. Will add the default implementation.

[the remark this answers is not in the document]
Comment (littlepig2013) [reply include/rocksdb/threadpool.h:60]: Maybe the upper-level function can take charge of detecting the inconsistency between the expected released threads and the real released threads? In our future use, we call `ReserveThreads` first and get the number of reserved threads, and when we release threads, we compare the returned one with the earlier returned reserved threads. Upper-level functions may want to release some number of threads, and also wonder how many threads are really released.

File: env/env_test.cc [hunk 4]
```diff
@@ -839,6 +839,46 @@ TEST_P(EnvPosixTestWithParam, DecreaseNumBgThreads) {
...
+  }
+  ASSERT_FALSE(tasks[0].TimedWaitUntilSleeping(kWaitMicros));
+  ASSERT_EQ(2U, env_->GetThreadPoolQueueLen(Env::Priority::HIGH));
+  ASSERT_TRUE(tasks[0].IsSleeping());
+  ASSERT_TRUE(!tasks[1].IsSleeping());
+  ASSERT_TRUE(!tasks[2].IsSleeping());
+
+  // Release 2 threads. Task 0, 1, 2 running; 1 reserved threads.
+  ASSERT_EQ(2, env_->ReleaseThreads(2, Env::Priority::HIGH));
+  Env::Default()->SleepForMicroseconds(kDelayMicros);
```
[the remark this answers is not in the document]
Comment (littlepig2013) [reply env/env_test.cc:863]: Yes. Since we have added the signal, I will remove this.

File: util/threadpool_imp.cc (hunk 2 above)
Comment (littlepig2013) [reply util/threadpool_imp.cc:370]: Yes. I will remove this.

File: util/threadpool_imp.cc (hunk 1 above)
Comment (littlepig2013) [reply util/threadpool_imp.cc:115]: Indeed, `num_waiting_` is definitely a better upper bound than `total_threads_limit_`. I will add it in the next commit.

File: include/rocksdb/threadpool.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (ajkr) [reply include/rocksdb/threadpool.h:60]: The question though is whether the upper-level will ever need to `ReleaseThreads()` with a larger `threads_to_be_released` than number of threads available to be released. If not, it should be good enough that the upper-level can learn how many threads were really released according to a returned `Status`:

- Return value `Status::OK`: exactly `threads_to_be_released` were released
- Return value `Status::InvalidArgument`: exactly zero threads were released

[the remark this answers is not in the document]
Comment (ajkr) [reply include/rocksdb/threadpool.h:60]: I think this API is OK with me for symmetry purposes, though. I think it's OK to treat this suggestion as optional even if the upper-level does not need the flexibility.

File: util/threadpool_imp.cc [hunk 5]
```diff
@@ -90,7 +90,8 @@ struct ThreadPoolImpl::Impl {
   }
 
   bool IsExcessiveThread(size_t thread_id) const {
-    return static_cast<int>(thread_id) >= total_threads_limit_;
+    return static_cast<int>(thread_id) + reserved_threads_ >=
+           total_threads_limit_;
```
Comment (ajkr) [inline util/threadpool_imp.cc:94]: This seems inconsistent with `HasExcessiveThread()`. Personally I'd not terminate/recreate thread-pool threads for reservations.

**Review by ajkr — APPROVED** [review]
LGTM

File: util/threadpool_imp.cc (hunk 5 above)
Comment (littlepig2013) [reply util/threadpool_imp.cc:94]: I see your concern. Will change it.

Comment (ajkr) [reply util/threadpool_imp.cc:94]: Oh good point. I guess we could resolve the inconsistency either by changing the name `IsExcessiveThread()` to something else like `IsBlockedThread()`, or by adding back the `reserved_threads_` into `HasExcessiveThread()`. Up to you.

File: util/threadpool_imp.cc [hunk 6]
```diff
@@ -100,6 +103,35 @@ struct ThreadPoolImpl::Impl {
...
+    std::unique_lock<std::mutex> lock(mu_);
+    int reserved_threads_in_success = threads_to_be_reserved;
+    // We can reserve at most num_waiting_ in total so the number of threads
+    // that can be reserved might be fewer than the desired one
+    if (reserved_threads_ + threads_to_be_reserved > num_waiting_) {
+      reserved_threads_in_success = num_waiting_ - reserved_threads_;
+      reserved_threads_ = num_waiting_;
+    } else {
+      reserved_threads_ += threads_to_be_reserved;
+    }
```
Comment (hx235) [inline util/threadpool_imp.cc:116]: Is it equivalent to 
```
reserved_threads_in_success = std::min(num_waiting_ - reserved_threads_, threads_to_be_reserved)
reserved_threads_ += reserved_threads_in_success;
```

Similar case for `ReleaseThreads()`

File: util/threadpool_imp.cc [hunk 7]
```diff
@@ -110,6 +142,10 @@ struct ThreadPoolImpl::Impl {
 
  int total_threads_limit_;
  std::atomic_uint queue_len_;  // Queue length. Used for stats reporting
+ // Number of reserved threads by other tasks in bgthreads_
+ int reserved_threads_;
+ // Number of waiting threads (Maximum number of threads that can be reserved)
+ int num_waiting_;
```
Comment (hx235) [inline util/threadpool_imp.cc:148]: nit: num_waiting_threads_

(I generally prefer more explicit naming for variable so that people don't have to look at comment to understand its intention)

File: util/threadpool_imp.cc [hunk 8]
```diff
@@ -110,6 +142,10 @@ struct ThreadPoolImpl::Impl {
 
  int total_threads_limit_;
  std::atomic_uint queue_len_;  // Queue length. Used for stats reporting
+ // Number of reserved threads by other tasks in bgthreads_
+ int reserved_threads_;
```
Comment (hx235) [inline util/threadpool_imp.cc:146]: nit: num_reserved_threads_ to be consistent with below

Comment (hx235) [reply util/threadpool_imp.cc:146]: "other tasks" sounds a bit random/unclear, although I understand it refers to tasks that reserve threads (e.g, subcompaction) as opposed to tasks do not reserve threads (e.g, flush) because I was exposed to this PR's motivation. 

Is there a way to better clarify the comment so that people w/o much context of your PR understand it?

File: util/threadpool_imp.cc (hunk 6 above)
Comment (littlepig2013) [reply util/threadpool_imp.cc:116]: Good catch! Will apply this change soon.

File: util/threadpool_imp.cc (hunk 8 above)
Comment (littlepig2013) [reply util/threadpool_imp.cc:146]: I see your point. I will revise it accordingly.

File: util/threadpool_imp.cc (hunk 7 above)
Comment (littlepig2013) [reply util/threadpool_imp.cc:148]: Sure.

File: util/threadpool_imp.cc [hunk 9]
```diff
@@ -189,10 +229,14 @@ void ThreadPoolImpl::Impl::BGThread(size_t thread_id) {
     // Wait until there is an item that is ready to run
     std::unique_lock<std::mutex> lock(mu_);
     // Stop waiting if the thread needs to do work or needs to terminate.
+    // increase num_waiting_ once this task has finished
```
Comment (hx235) [inline util/threadpool_imp.cc:232]: Should this be "increase num_waiting_ once the thread starts waiting" to be consistent with "decrease num_waiting_ once the thread is not waiting" below?

nit: Start with a capitalized letter in the comments to be consistent with others such as  "Increase ..." or "Decrease ..." (below)

Comment (littlepig2013) [reply util/threadpool_imp.cc:232]: That was my typo. Thanks for pointing out this.

File: util/threadpool_imp.h [hunk 10]
```diff
@@ -33,7 +33,9 @@ class ThreadPoolImpl : public ThreadPool {
   void JoinAllThreads() override;
 
   // Set the number of background threads that will be executing the
-  // scheduled jobs.
+  // scheduled jobs. (This function will reset reserved_threads to 0,
+  // reset is only triggered when total_threads_limit changes, typically,
+  // (num > total_threads_limit) || (num < total_threads_limit && allow_reduce)
```
Comment (hx235) [inline util/threadpool_imp.h:38]: Comments not needed?

Comment (littlepig2013) [reply util/threadpool_imp.h:38]: Ohh, thanks a lot for catching this! This was an old comment for my earlier implementation. I will remove this.

File: util/threadpool_imp.cc [hunk 11]
```diff
@@ -100,6 +103,35 @@ struct ThreadPoolImpl::Impl {
...
+    int reserved_threads_in_success = threads_to_be_reserved;
+    // We can reserve at most num_waiting_ in total so the number of threads
+    // that can be reserved might be fewer than the desired one
+    if (reserved_threads_ + threads_to_be_reserved > num_waiting_) {
+      reserved_threads_in_success = num_waiting_ - reserved_threads_;
+      reserved_threads_ = num_waiting_;
+    } else {
+      reserved_threads_ += threads_to_be_reserved;
+    }
+    WakeUpAllThreads();
```
Comment (hx235) [inline util/threadpool_imp.cc:117]: I'm not following why do we need to `WakeUpAllThreads();` at reserve - can I have more explanation about this in PR conversation?

Comment (littlepig2013) [reply util/threadpool_imp.cc:117]: Sure. As you can see, there is a remaining bug that makes many CI tests failed. I am fixing that bug and I think `WakeUpAllThreads()` will be needed in `ReserveThreads` to fix that bug. I will tell you later after I finalize the solution.

**Comment by hx235** [comment]
As it's very hard to prove correctness of concurrent programming by testing, I'd like to ponder the current concurrent design in this PR a bit more and check my understanding (so plz correct me if I'm wrong).

Below is my visual understanding of thread pool composition after this PR: 
threadpool_composition

My conceptual understanding:
(1) The `Reserve()` and `Release()` functions should only affect the counter `num_reserved_threads` (i.e, [A1])  and not `num_waiting_threads`. The reason is that the "reserved threads" are executed outside of this Env thread pool 
(2) As a consequence of (1),  `# available threads for reservation or running w/o reservation` (i.e, [A2]) will decrease after `Reserve()` and increase after `Release()`
(3) The point of waking up threads is to signal them that some conditions determining whether a thread can run or not has changed and the sleeping threads should wake up and check these conditions again in case some gets to run. 
As a result of (2), such conditions change. However, only when `# available threads for reservation or running w/o reservation` increases, such condition is a promising condition for threads to run. That's why I left a question above for further explanation on why we need `WakeUpAllThreads();` in `Reserve()`.

File: util/threadpool_imp.cc (hunk 11 above)
Comment (hx235) [reply util/threadpool_imp.cc:117]: See below for more context behind this question.

File: util/threadpool_imp.cc (hunk 5 above)
Comment (hx235) [reply util/threadpool_imp.cc:96]: I am not quite following the why the new `IsBlockedThread()` implementation is correct in preventing extra threads being woken up as a result of not considering reserved threads. See (4) in my review summary https://github.com/facebook/rocksdb/pull/10278#issuecomment-1175764224 for more.

**Comment by littlepig2013** [comment]
> 

You are absolutely correct! That is the way I'm fixing it! And you see why we need to add `WakeUpAllThreads` at reserve now? Because after we call the reserve function, some threads are supposed to be blocked no matter how its thread id relate to the bgthreads_.size()! That means, if some threads are not excessive, they still need to be blocked!

**Comment by littlepig2013** [comment]
> As it's very hard to prove correctness of concurrent programming by testing, I'd like to ponder the current concurrent design in this PR a bit more and check my understanding (so plz correct me if I'm wrong).
> 
> Below is my visual understanding of thread pool composition after this PR: threadpool_composition
> 
> My conceptual understanding: (1) The `Reserve()` and `Release()` functions should only affect the counter `num_reserved_threads` (i.e, [A1]) and not `num_waiting_threads`. The reason is that the "reserved threads" are executed outside of this Env thread pool "
> 
> (2) As a consequence of (1), `# available threads for reservation or running w/o reservation` (i.e, [A2]) will decrease after `Reserve()` and increase after `Release()`
> 
> (3) The point of waking up threads is to signal them that some conditions determining whether a thread can run or not has changed and the sleeping threads should wake up and check these conditions again in case some gets to run. As a result of (2), such conditions change. However, only when `# available threads for reservation or running w/o reservation` increases, such condition is a promising condition for threads to run. That's why I left a question above for further explanation on why we need `WakeUpAllThreads();` in `Reserve()`.
> 
> (4) When it comes to actually "check these condition", should we keep `IsExcessiveThread` as it is but further check ``# available threads for reservation or running w/o reservation > 0`that is` num_waiting_threads - num_reserved_threads > 0`? The check intuitively makes sense to me as after passing the `IsExcessiveThread` due to, we need to consider an extra layer of constraint imposed by this PR, threads reservation.
> 
> I don't see this being explicitly checked in current impl and I am not able to understand why `return static_cast<int>(thread_id) + reserved_threads_ >= total_threads_limit_;` is enough. So I will need a bit help in understanding this.

The earlier implementation tries to implement the reservation through controling the bgthread size. And as now we move to another implementation (keeping bgthreads size), the mixed implementation makes the logic a big buggy.

**Comment by hx235** [comment]
> That means, if some threads are not excessive, they still need to be blocked!

What is the reason on why they still need to be blocked? Sorry I still don't see why we need the wake up in Release().... 
Can we add another check `(num_waiting_threads - num_reserved_threads > 0)` in the while condition and eliminate that wakeup?

**Comment by littlepig2013** [comment]
> > That means, if some threads are not excessive, they still need to be blocked!
> 
> What is the reason on why they still need to be blocked - is it because of the threads reservation? Even though, I still don't see why we need the wake up in Release(). Can we add another check `(num_waiting_threads - num_reserved_threads > 0)` in the while condition and eliminate that wakeup?

The short answer is to block threads with a correct condition. Before reservation, some threads are blocked because the queue is empty or it is an excessive thread. After reservation, some threads are blocked because they are reserved. At this time, even if some tasks are submitted, they will be put into the queue instead of being executed.  

> > That means, if some threads are not excessive, they still need to be blocked!
> 
> What is the reason on why they still need to be blocked - is it because of the threads reservation? Even though, I still don't see why we need the wake up in Release(). Can we add another check `(num_waiting_threads - num_reserved_threads > 0)` in the while condition and eliminate that wakeup?

File: util/threadpool_imp.cc [hunk 12]
```diff
@@ -189,10 +220,20 @@ void ThreadPoolImpl::Impl::BGThread(size_t thread_id) {
...
+    num_waiting_threads_++;
+
+    // When not exist_all_threads and the current thread id is not the last
+    // excessive thread, it may be blocked due to 3 reasons: 1) queue is empty
+    // 2) it is the excessive thread (not the last one)
+    // 3) the number of reserved threads reaches the number of waiting threads
     while (!exit_all_threads_ && !IsLastExcessiveThread(thread_id) &&
-           (queue_.empty() || IsExcessiveThread(thread_id))) {
+           (queue_.empty() || IsExcessiveThread(thread_id) ||
+            num_waiting_threads_ == reserved_threads_)) {
```
Comment (ajkr) [inline util/threadpool_imp.cc:232]: Sorry I'm missing the intuition. How does `num_waiting_threads_` ever reach `reserved_threads_` when queue is nonempty and the thread is not excessive? Won't it spin the CPU?

Comment (littlepig2013) [reply util/threadpool_imp.cc:232]: The queue can be nonempty after reservation. Several tasks are then put in the queue. Then A thread releases fewer number of threads, at this moment, a third thread issues another reservation.

Comment (littlepig2013) [reply util/threadpool_imp.cc:232]: Typically, earlier buggy implementation failed at `ASSERT_EQ(1U, env_->GetThreadPoolQueueLen(Env::Priority::HIGH));' Because at that point, thread with id 0 has been waiting and its id is not excessive, this will make thread 0 execute the incoming task 3, instead of being blocked. Adding the `num_waiting_thread_ == reserved_threads_` can ensure thread 0 is blocked and Task 3 is in the queue.

Comment (littlepig2013) [reply util/threadpool_imp.cc]: updated with replacing the condition with `num_waiting_threads_  <= reserved_threads_`.

Comment (hx235) [reply util/threadpool_imp.cc]: Will need to catch up on this tomorrow but I am thinking (1) whether num_waiting_threads_  < num_reserved_threads is legitimate and (2) if it is, whether we should revise the design to eliminate that cuz it’s pretty counter-intuitive given the Reserve() algorithm we are pursuing.

Comment (littlepig2013) [reply util/threadpool_imp.cc]: For the current design, it is very possible that num_waiting_threads < num_reserved_threads because of the excessive case (andrew mentioned earlier). I will replace num_waiting_threads - num_reserved_threads with max(num_waiting_threads - num_reserved_threads, 0) in the next commit

File: util/thread_list_test.cc [hunk 13]
```diff
@@ -126,9 +126,11 @@ TEST_F(ThreadListTest, SimpleColumnFamilyInfoTest) {
   const int kLowPriorityThreads = 5;
   const int kSimulatedHighPriThreads = kHighPriorityThreads - 1;
   const int kSimulatedLowPriThreads = kLowPriorityThreads / 3;
+  const int kDelayMicros = 1000000;
   env->SetBackgroundThreads(kHighPriorityThreads, Env::HIGH);
   env->SetBackgroundThreads(kLowPriorityThreads, Env::LOW);
-
+  // Wait 1 second so that threads start
+  Env::Default()->SleepForMicroseconds(kDelayMicros);
```
Comment (hx235) [inline util/thread_list_test.cc:133]: Is there any way to assert threads are started after this delay IF it's needed to proceed? Fix-sized sleep almost always concern me on flakiness (i.e, what if threads need more than 1 sec to start in some rare occasion)?

Comment (hx235) [reply util/thread_list_test.cc:133]: Same for the unit test above

File: env/env_test.cc [hunk 14]
```diff
@@ -839,6 +839,64 @@ TEST_P(EnvPosixTestWithParam, DecreaseNumBgThreads) {
...
+  constexpr int kWaitMicros = 10000000;  // 10seconds
+  std::vector<test::SleepingBackgroundTask> tasks(4);
+  // Set number of thread to 4 first.
+  env_->SetBackgroundThreads(4, Env::Priority::HIGH);
+  // Wait until all threads start (num_waiting accumulates)
+  // Reserve 3 threads
+  Env::Default()->SleepForMicroseconds(kDelayMicros);
+  ASSERT_EQ(3, env_->ReserveThreads(3, Env::Priority::HIGH));
+
+  // Schedule 3 tasks. 0 running; Task 1, 2 waiting; 3 reserved threads.
```
Comment (hx235) [inline env/env_test.cc:852]: nit: "Task 0 running; .."

File: env/env_test.cc [hunk 15]
```diff
@@ -839,6 +839,64 @@ TEST_P(EnvPosixTestWithParam, DecreaseNumBgThreads) {
...
+  ASSERT_FALSE(tasks[1].TimedWaitUntilSleeping(kWaitMicros));
+  ASSERT_FALSE(tasks[2].TimedWaitUntilSleeping(kWaitMicros));
+  ASSERT_EQ(0U, env_->GetThreadPoolQueueLen(Env::Priority::HIGH));
+  ASSERT_TRUE(tasks[1].IsSleeping());
+  ASSERT_TRUE(tasks[2].IsSleeping());
+
+  // Task 1, 2 running; 1 reserved thread.
+  tasks[0].WakeUp();
+  ASSERT_FALSE(tasks[0].TimedWaitUntilDone(kWaitMicros));
+  Env::Default()->SleepForMicroseconds(kDelayMicros);
```
Comment (hx235) [inline env/env_test.cc:874]: Should we add `ASSERT_EQ(0, env_->ReserveThreads(3, Env::Priority::HIGH));` as a prior before any change that can lead to  `ASSERT_EQ(1, env_->ReserveThreads(3, Env::Priority::HIGH));` below to prove the point of `1 more thread can be reserved`?

File: util/thread_list_test.cc (hunk 13 above)
Comment (littlepig2013) [reply util/thread_list_test.cc:133]: I did not change the above unit test. The above one is for the origin threadpool. I will think about how to add the sync point.

File: env/env_test.cc (hunk 14 above)
Comment (littlepig2013) [reply env/env_test.cc:852]: yeah, will do!

File: env/env_test.cc (hunk 15 above)
Comment (littlepig2013) [reply env/env_test.cc:874]: We have it at line 850.

Comment (littlepig2013) [reply env/env_test.cc:874]: No problem, I will add it.

File: env/env_test.cc [hunk 16]
```diff
@@ -839,6 +839,64 @@ TEST_P(EnvPosixTestWithParam, DecreaseNumBgThreads) {
   WaitThreadPoolsEmpty();
 }
 
+TEST_P(EnvPosixTestWithParam, ReserveThreads) {
```
[the remark this answers is not in the document]
Comment (hx235) [reply env/env_test.cc:842]: One more scenario I'd like to see in testing is when `SetBackgroundThreads()` is set to a smaller number in the middle (e.g, shrink the thread pool) and see if woken up threads respect existing reservation. Basically it's the test case to verify the scenario Andrew mentioned.

File: util/threadpool_imp.cc [hunk 17]
```diff
@@ -110,6 +134,13 @@ struct ThreadPoolImpl::Impl {
 
  int total_threads_limit_;
  std::atomic_uint queue_len_;  // Queue length. Used for stats reporting
+ // Number of reserved threads, managed by ReserveThreads(..) and
+ // ReleaseThreads(..), if a thread id is beyond total_threads_limit_ -
+ // reserved_threads_, its thread will be blocked to ensure the reservation
+ // mechanism
```
Comment (hx235) [inline util/threadpool_imp.cc:140]: Comments needed update as now we are not doing "if a thread id is beyond total_threads_limit_ // reserved_threads_,"?

File: util/threadpool_imp.cc [hunk 18]
```diff
@@ -110,6 +134,13 @@ struct ThreadPoolImpl::Impl {
...
+ int reserved_threads_;
+ // Number of waiting threads (Maximum number of threads that can be reserved)
```
Comment (hx235) [inline util/threadpool_imp.cc:142]: I suggest adding the " in rare cases, num_waiting_threads_ could be less than reserved_threads
due to SetBackgroundThreadInternal or last excessive threads." comment here instead so future developer can be aware of this just by looking at these two variables declaration.

File: util/threadpool_imp.cc [hunk 19]
```diff
@@ -189,10 +224,20 @@ void ThreadPoolImpl::Impl::BGThread(size_t thread_id) {
     // Wait until there is an item that is ready to run
     std::unique_lock<std::mutex> lock(mu_);
     // Stop waiting if the thread needs to do work or needs to terminate.
+    // Increase num_waiting_threads_ once this task has started waiting
+    num_waiting_threads_++;
+
+    // When not exist_all_threads and the current thread id is not the last
+    // excessive thread, it may be blocked due to 3 reasons: 1) queue is empty
+    // 2) it is the excessive thread (not the last one)
+    // 3) the number of reserved threads is not greater than waiting threads
```
Comment (hx235) [inline util/threadpool_imp.cc:233]: "...the number of **waiting** threads is not greater than **reserved** threads"

**Review by hx235 — CHANGES_REQUESTED** [review]
Thanks for the progress so far!!! I left suggestions mainly in UT and comments. Sorry for being extra picky on these details for the following reasons:

(1) I still believe concurrent programming is very hard to prove by testing. When I'd imagine future developers come in without any first/second-hand experience with your PR, a good UT and comments help them reason about the program a lot. Your latest version's comment has already been a big step and let us push forward a bit more 😉! 

(I have a great bias toward this cuz https://github.com/facebook/rocksdb/pull/8596 reminds me of my own painful experience to stare at less well-documented concurrent codes and god knows how many chocolates I ate during those sleepless nights ...)

(2) Flaky tests even at its simple form can take someone without first/second-hand experience on the original PR a while to understand what's going on (along with probably a lot of chocolates at least for me ..)

File: util/threadpool_imp.cc (hunk 17 above)
Comment (littlepig2013) [reply util/threadpool_imp.cc:140]: Sure. Will update it.

File: util/threadpool_imp.cc (hunk 12 above)
Comment (hx235) [reply util/threadpool_imp.cc]: Right - `SetBackgroundThreadsInternal` can be called at any point and we can't terminate any running threads or reserved threads but num_waiting_threads.

It's make things less obvious cuz people or at least I'd easily think "We just do now allow more threads to be reserved when that occurs." should NOT happen (cuz reserved threads are like running threads, just running elsewhere). I might think of failing or delaying `SetBackgroundThreadsInternal` in this case. 

However, this might be too big a follow-up work for too infrequent use case.  So I'm good with adding a brief note and explanation
about "why there would ever be num_waiting_threads < num_reserved_threads" around the `num_waiting_threads` declaration" around the variable "num_waiting_threads" declaration as I suggested in another comment.

File: util/threadpool_imp.cc (hunk 18 above)
Comment (littlepig2013) [reply util/threadpool_imp.cc:142]: Yeah, will add it.

File: util/threadpool_imp.cc (hunk 12 above)
Comment (littlepig2013) [reply util/threadpool_imp.cc]: As you can see from the origin implementation, in some cases, the number of running threads could be larger than the thread pool. In other words, the total_threads_limit_ is not a strict constraint, and that's why we allow that some extra reserved threads to run even though it is larger than the number of waiting threads.

File: env/env_test.cc (hunk 16 above)
[the remark this answers is not in the document]
Comment (littlepig2013) [reply env/env_test.cc:842]: Yes, adding that scenario is definitely better. I will think harder to design this case.

File: env/env_test.cc (hunk 15 above)
Comment (littlepig2013) [reply env/env_test.cc]: Yes, will add it.

File: util/threadpool_imp.cc [hunk 20]
```diff
@@ -155,6 +191,8 @@ void ThreadPoolImpl::Impl::JoinThreads(bool wait_for_jobs_to_complete) {
   // prevent threads from being recreated right after they're joined, in case
   // the user is concurrently submitting jobs.
   total_threads_limit_ = 0;
+  reserved_threads_ = 0;
```
Comment (hx235) [inline util/threadpool_imp.cc:194]: No need to take action - I just want to note that it's interesting we are setting reserved_threads_ = 0 while ThreadPoolImpl does not actually have the power to terminate those reserved threads. IMO, this is the biggest drawback of this approach exchanged for the benefit of implementational simplicity.

File: util/threadpool_imp.cc (hunk 12 above)
Comment (hx235) [reply util/threadpool_imp.cc]: Got you!

File: env/env_test.cc [hunk 21]
```diff
@@ -1271,7 +1401,6 @@ TEST_P(EnvPosixTestWithParam, MultiRead) {
             }
           }
         });
-    ROCKSDB_NAMESPACE::SyncPoint::GetInstance()->EnableProcessing();
```
Comment (hx235) [inline env/env_test.cc:1274]: unrelated change?

Comment (littlepig2013) [reply env/env_test.cc:1274]: It's a bit weird that i did not change that.

File: env/env_test.cc [hunk 22]
```diff
@@ -839,6 +839,136 @@ TEST_P(EnvPosixTestWithParam, DecreaseNumBgThreads) {
...
+  constexpr int kWaitMicros = 10000000;  // 10seconds
+  std::vector<test::SleepingBackgroundTask> tasks(4);
+  ROCKSDB_NAMESPACE::SyncPoint::GetInstance()->EnableProcessing();
+  // Set the sync point to ensure thread 0 can terminate
+  ROCKSDB_NAMESPACE::SyncPoint::GetInstance()->LoadDependency(
+      {{"ThreadPoolImpl::BGThread::Termination:th0",
+        "EnvTest::ReserveThreads:0"}});
+  // Empty the thread pool to ensure all the threads can start later
+  env_->SetBackgroundThreads(0, Env::Priority::HIGH);
+  ASSERT_EQ(env_->GetBackgroundThreads(Env::HIGH), 0);
```
Comment (hx235) [inline env/env_test.cc:856]: So it seems to me that it's possible for `ASSERT_EQ(env_->GetBackgroundThreads(Env::HIGH), 0);` to be true while `ThreadPoolImpl::BGThread::Termination:th0` has not been run?

Comment (littlepig2013) [reply env/env_test.cc:856]: i think we are safe to remove it

File: env/env_test.cc [hunk 23]
```diff
@@ -839,6 +839,136 @@ TEST_P(EnvPosixTestWithParam, DecreaseNumBgThreads) {
...
+  ASSERT_EQ(0U, env_->GetThreadPoolQueueLen(Env::Priority::HIGH));
+
+  // At most 1 thread can be released
+  ASSERT_GT(2, env_->ReleaseThreads(3, Env::Priority::HIGH));
+  tasks[2].WakeUp();
+  ASSERT_FALSE(tasks[2].TimedWaitUntilDone(kWaitMicros));
+  tasks[3].WakeUp();
+  ASSERT_FALSE(tasks[3].TimedWaitUntilDone(kWaitMicros));
+  WaitThreadPoolsEmpty();
+  ROCKSDB_NAMESPACE::SyncPoint::GetInstance()->DisableProcessing();
```
Comment (hx235) [inline env/env_test.cc:969]: Are these tests scenarios that haven't been covered previously OR just clean-up? Just want to make sure we are also not making the test unnecessarily long due to duplicating testing scenarios.  

Not really sure what `0/1 reserved threads` means - is it "0 or 1 reserved thread"?

Comment (littlepig2013) [reply env/env_test.cc:969]: Line 964 - 968 is just clean-up. 

yes, I will change it to make it more clear.

**Review by hx235 — APPROVED** [review]
Thanks for your efforts to cover different scenarios in the UT along with great comments! Left a few nits!
